src/miner.cpp

```diff
@@ -138,82 +276,76 @@ CBlockTemplate* CreateNewBlock(const CScript& scriptPubKeyIn)
+            CAmount& nTxFee = txinfo.nTxFee;
             CAmount nTotalIn = 0;
-            bool fMissingInputs = false;
             BOOST_FOREACH(const CTxIn& txin, tx.vin)
             {
                 // Read prev transaction
-                if (!view.HaveCoins(txin.prevout.hash))
+                CAmount nValueIn;
+                int nConf;
+                if (view.HaveCoins(txin.prevout.hash))
                 {
-                    // This should never happen; all transactions in the memory
-                    // pool should connect to either transactions in the chain
-                    // or other transactions in the memory pool.
-                    if (!mempool.mapTx.count(txin.prevout.hash))
-                    {
-                        LogPrintf("ERROR: mempool transaction missing input\n");
-                        if (fDebug) assert("mempool transaction missing input" == 0);
-                        fMissingInputs = true;
-                        if (porphan)
-                            vOrphan.pop_back();
-                        break;
-                    }
-
-                    // Has to wait for dependencies
-                    if (!porphan)
-                    {
-                        // Use list for automatic deletion
-                        vOrphan.push_back(COrphan(&tx));
-                        porphan = &vOrphan.back();
-                    }
-                    mapDependers[txin.prevout.hash].push_back(porphan);
-                    porphan->setDependsOn.insert(txin.prevout.hash);
-                    nTotalIn += mempool.mapTx[txin.prevout.hash].GetTx().vout[txin.prevout.n].nValue;
-                    continue;
+                    const CCoins* coins = view.AccessCoins(txin.prevout.hash);
+                    assert(coins);
+                    // Input is confirmed
+                    nConf = nHeight - coins->nHeight;
+                    nValueIn = coins->vout[txin.prevout.n].nValue;
+                    dPriority += (double)nValueIn * nConf;
+                }
+                else
+                if (mempool.mapTx.count(txin.prevout.hash))
+                {
+                    // Input is still unconfirmed
+                    const uint256& hashPrev = txin.prevout.hash;
+                    nValueIn = mempool.mapTx[hashPrev].GetTx().vout[txin.prevout.n].nValue;
+                    txinfo.addDependsOn(hashPrev);
+                    mapInfoById[hashPrev].setDependents.insert(hash);
+                    nConf = 0;
+                }
+                else
+                {
+                    // We don't know where the input is
+                    // In this case, it's impossible to include this transaction in a block, so mark it invalid and move on
+                    txinfo.fInvalid = true;
+                    LogPrintf("priority %s invalid input %s\n", txinfo.hash.ToString().substr(0,10).c_str(), txin.prevout.hash.ToString().substr(0,10).c_str());
+                    goto nexttxn;
```

Comment:
What about replacing this part

```
                    txinfo.fInvalid = true;
                    goto nexttxn;
                 }
                  nTotalIn += nValueIn;
             }
             nTxFee = nTotalIn - tx.GetValueOut(); 
             mempool.ApplyDeltas(hash, dPriority, nTotalIn);
             vecPriority.push_back(&txinfo); 
nexttxn:    (void)1;
```

with the following?

```
                    txinfo.fInvalid = true;
                 }
                 if (!txinfo.fInvalid)
                    nTotalIn += nValueIn;
             }
             if (!txinfo.fInvalid) {
                 nTxFee = nTotalIn - tx.GetValueOut(); 
                 mempool.ApplyDeltas(hash, dPriority, nTotalIn);
                 vecPriority.push_back(&txinfo);
             }
```

https://xkcd.com/292/